components/debugger/lib.rs

```diff
@@ -2,49 +2,68 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#[macro_use]
+extern crate log;
 extern crate util;
-extern crate websocket;
+extern crate ws;
 
+use std::sync::mpsc;
+use std::sync::mpsc::channel;
 use util::thread::spawn_named;
-use websocket::{Message, Receiver, Sender, Server};
-use websocket::message::Type;
+use ws::{Builder, CloseCode, Handler, Handshake};
 
-pub fn start_server(port: u16) {
-    println!("Starting debugger server.");
-    spawn_named("debugger-server".to_owned(), move || {
-        run_server(port)
-    });
+enum Message {
+    ShutdownServer,
+}
+
+pub struct Sender(mpsc::Sender<Message>);
+
+struct Connection {
+    sender: ws::Sender
+}
+
+impl Handler for Connection {
+    fn on_open(&mut self, _: Handshake) -> ws::Result<()> {
+        println!("Connection opened.");
+        Ok(())
+    }
+
+    fn on_close(&mut self, _: CloseCode, _: &str) {
+        println!("Connection closed.");
+    }
+
+    fn on_message(&mut self, message: ws::Message) -> ws::Result<()> {
+        self.sender.send(message)
+    }
 }
 
-fn run_server(port: u16) {
-    let server = Server::bind(("127.0.0.1", port)).unwrap();
-    for connection in server {
-        spawn_named("debugger-connection".to_owned(), move || {
-            let connection = connection.unwrap();
-            let request = connection.read_request().unwrap();
-            let response = request.accept();
-            let client = response.send().unwrap();
-            let (mut sender, mut receiver) = client.split();
-            for message in receiver.incoming_messages() {
-                let message: Message = message.unwrap();
-                match message.opcode {
-                    Type::Close => {
-                        let message = Message::close();
-                        sender.send_message(&message).unwrap();
-                        break;
-                    }
-                    Type::Ping => {
-                        let message = Message::pong(message.payload);
-                        sender.send_message(&message).unwrap();
-                    }
-                    Type::Text => {
-                        sender.send_message(&message).unwrap();
-                    }
-                    _ => {
-                        panic!("Unexpected message type.");
-                    }
+pub fn start_server(port: u16) -> Sender {
+    println!("Starting server.");
+    let (sender, receiver) = channel();
+    spawn_named("debugger".to_owned(), move || {
+        let socket = Builder::new().build(|sender: ws::Sender| {
+            Connection { sender: sender }
+        }).unwrap();
+        let sender = socket.broadcaster();
+        spawn_named("debugger-websocket".to_owned(), move || {
```

Comment:
It dies when the main debugger thread receives a ShutdownServer message, breaks out of its main message loop, and calls sender.shutdown().unwrap(). This will break the websocket server out of its listen loop, allowing the thread to clean itself up.